### PR TITLE
HMAC authentication

### PR DESCRIPTION
**Pull request by jacobgreenleaf**

### CODE REVIEW

File: server/server.go [hunk 1]
```diff
@@ -161,61 +230,116 @@ func (s *Server) uploadFile(uploadFile io.Reader, w http.ResponseWriter, fileNam
...
+			default:
+				break
+			}
+
+			resp.Write(w)
 		}
+	}
 
-		s.uploadFile(uploadFile, w, "", thumbs)
+	authenticatedEndpoint := func(endpoint uploadEndpoint, extractor fileExtractor) http.HandlerFunc {
```
Comment (briankassouf) [inline server/server.go]: `var authenticatedEndpoint uploadEndpoint`

File: README.md [hunk 2]
```diff
@@ -131,3 +139,58 @@ curl -i http://127.0.0.1:8080/file \
...
+```javascript
+{"data":{"link":"/tmp/original/J/a/Jafq9IH","mime":"image/jpeg","name":"s9zxmYe.jpg","hash":"Jafq9IH","size":81881,"width":380,"height":430,"ocrtext":"change\np.roject .\n\n  \n  \n\n  forg@ot to git p.ull before\n- .-+#~+):,-r,ad)q..,i,ng so/ /","thumbs":{},"user_id":"\u0001"},"status":200,"success":true}
+```
+
+#### Request to other user's account:
+
+```
+curl -i http://127.0.0.1:8080/user/2/url \
+-d 'image=http://i.imgur.com/s9zxmYe.jpg' \
+-H 'Authorization: {"user_id":1,"grant_time":"2010-06-01T00:00:00Z","grant_duration_sec":31536000}' \
```
Comment (briankassouf) [inline README.md]: Should this be a different user_id?

Comment (jacobgreenleaf) [inline README.md]: no, this is demonstrating that you can't upload to others' account.

Comment (briankassouf) [inline README.md]: Ah okay, got it

File: server/server.go (hunk 1 above)
Comment (briankassouf) [inline server/server.go]: @jacobgreenleaf Want to make this change then we can merge?

Comment (jacobgreenleaf) [inline server/server.go]: I'm gonna add another test to get coverage on `github.com/Imgur/mandible/server/server.go:88:		NewAuthenticatedServer		0.0%`.

File: Dockerfile [hunk 3]
```diff
@@ -0,0 +1,16 @@
+FROM golang:1.4
+ENV PORT 80
+EXPOSE 80
+RUN apt-get update && apt-get install -yqq aspell aspell-en libaspell-dev tesseract-ocr tesseract-ocr-eng graphicsmagick graphicsmagick-imagemagick-compat optipng exiftool libjpeg-progs
+RUN wget https://github.com/johnlinp/meme-ocr/raw/master/tessdata/joh.traineddata -O /usr/share/tesseract-ocr/tessdata/meme.traineddata
+RUN mkdir -p /etc/mandible /tmp/imagestore
+ADD docker/conf.json /etc/mandible/conf.json
+ENV MANDIBLE_CONF /etc/mandible/conf.json
```
Comment (briankassouf) [inline Dockerfile]: :+1:

**Pull request merged**